src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -46,29 +57,77 @@ uint16_t GenericOTARequestorDriver::GetMaxDownloadBlockSize()
     return 1024;
 }
 
-void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error)
+void StartDelayTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    static_cast<GenericOTARequestorDriver *>(appState)->DriverSendQuery();
+}
+
+bool ProviderLocationsEqual(const ProviderLocation::Type & a, const ProviderLocation::Type & b)
 {
-    // TODO: Schedule the next QueryImage
+    if ((a.fabricIndex == b.fabricIndex) && (a.providerNodeID == b.providerNodeID) && (a.endpoint == b.endpoint))
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error) {}
+
+void GenericOTARequestorDriver::HandleIdleState()
+{
+    // Default profivers timer runs if and only if the OTARequestor's update state is kIdle.
+    // Must (re)start the timer every time we enter the kIdle state
+    StartDefaultProviderTimer();
 }
 
 void GenericOTARequestorDriver::UpdateAvailable(const UpdateDescription & update, System::Clock::Seconds32 delay)
 {
+    // IMPLEMENTATION CHOICE:
+    // This implementation unconditionally downloads an available update
+
     VerifyOrDie(mRequestor != nullptr);
-    ScheduleDelayedAction(UpdateFailureState::kDownloading, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); });
+    ScheduleDelayedAction(
+        delay, [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); }, this);
 }
 
 void GenericOTARequestorDriver::UpdateNotFound(UpdateNotFoundReason reason, System::Clock::Seconds32 delay)
 {
     VerifyOrDie(mRequestor != nullptr);
 
+    ProviderLocation::Type providerLocation;
+
+    switch (reason)
+    {
+    case UpdateNotFoundReason::UpToDate:
+        return;
+        break;
+    case UpdateNotFoundReason::Busy:
+        // This will schedule a query with the same provider
+        break;
+    case UpdateNotFoundReason::ConnectionFailed:
+    case UpdateNotFoundReason::NotAvailable:
+        // IMPLEMENTATION CHOICE:
+        // This implementation schedules a query only if a different provider is available
+        if ((DetermineProviderLocation(providerLocation) != true) || ProviderLocationsEqual(providerLocation, mLastUsedProvider))
```

Comment:
What if we got `NotAvailable`as a result of AnnounceOTAProvider command? On a retry, we would be selecting from the default OTA list and not retrying what was in the announce command

Reply:
Yep. 

> Handling NotAvailable value in Status field
The NotAvailable status indicates that there is definitely no update available.
The OTA Requestor MAY choose to attempt a QueryImage command on a different OTA Provider in
its OTA Provider List, in which case the timing for the query SHALL follow the query timing
constraints expressed in the previous paragraphs of this section.